stan/math/prim/fun/lub_constrain.hpp

```diff
@@ -97,27 +84,22 @@ inline return_type_t<T, L, U> lub_constrain(const T& x, const L& lb,
  *   the free scalar.
  * @throw std::domain_error if ub <= lb
  */
-template <typename T, typename L, typename U, typename S>
-inline return_type_t<T, L, U> lub_constrain(const T& x, const L& lb,
-                                            const U& ub, S& lp) {
-  using std::exp;
-  using std::log;
-  check_less("lub_constrain", "lb", lb, ub);
-  if (lb == NEGATIVE_INFTY) {
-    return ub_constrain(x, ub, lp);
-  }
-  if (ub == INFTY) {
-    return lb_constrain(x, lb, lp);
-  }
-  T inv_logit_x = inv_logit(x);
-  if (x > 0) {
-    T exp_minus_x = exp(-x);
-    lp += log(ub - lb) - x - 2 * log1p(exp_minus_x);
-  } else {
-    T exp_x = exp(x);
-    lp += log(ub - lb) + x - 2 * log1p(exp_x);
-  }
-  return fma(ub - lb, inv_logit_x, lb);
+template <typename T, typename L, typename U>
+inline auto lub_constrain(const T& x, const L& lb, const U& ub,
+                          return_type_t<T, L, U>& lp) {
+  const auto& x_ref = to_ref(x);
+  const auto& lb_ref = to_ref(lb);
+  const auto& ub_ref = to_ref(ub);
+
+  check_less("lub_constrain", "lb", value_of(lb_ref), value_of(ub_ref));
+  check_finite("lub_constrain", "lb", value_of(lb_ref));
+  check_finite("lub_constrain", "ub", value_of(ub_ref));
+
+  const auto& diff = to_ref(subtract(ub_ref, lb_ref));
+  
+  lp += sum(add(log(diff),
+                subtract(-abs(x_ref), multiply(2, log1p_exp(-abs(x_ref))))));
```

Comment:
I added `abs` calls here to symmetrize the function around zero so that the finite differences would give zero. I kinda don't like it cause it depends on `abs` returning non-nan gradients at zero. I'd honestly be happer just upping the tolerances in that Stan tests.

The other option is to write out the custom autodiff here. It's just annoying to vectorize that conditional (look at the code we're replacing to see what is happening -- `x - 2 * log1p(exp(x))` is equal to `-x - 2 * log1p(exp(-x))`).

Reply:
Hmm, I think I'm okay with upping the tolerances since this is a finite diff issue more than an autodiff issue